demo-app/src/main/java/io/opentelemetry/android/demo/clients/RecommendationService.kt

```diff
@@ -0,0 +1,29 @@
+package io.opentelemetry.android.demo.clients
+
+import io.opentelemetry.android.demo.model.Product
+import io.opentelemetry.android.demo.ui.shop.cart.CartViewModel
+
+class RecommendationService(
+    private val productCatalogClient: ProductCatalogClient,
+    private val cartViewModel: CartViewModel
+) {
+
+    fun getRecommendedProducts(currentProduct: Product): List<Product> {
+        val allProducts = productCatalogClient.get()
+        val cartItems = cartViewModel.cartItems.value
+
+        return allProducts.filter { product ->
+            product.id != currentProduct.id && cartItems.none { it.product.id == product.id }
+        }.shuffled().take(4)
+    }
+
+    fun getRecommendedProducts(): List<Product> {
+        val allProducts = productCatalogClient.get()
+        val cartItems = cartViewModel.cartItems.value
+
+        return allProducts.filter { product ->
+            cartItems.none { it.product.id == product.id }
+        }.shuffled().take(4)
+    }
+}
```

Comment:
I think there's a small opportunity to improve this by factoring out a method called something like `getAllNonCartProducts()`. Just like it says in the name, that method would return all products from the catalog that are not currently in the cart. Both existing methods here can then delegate to that new method.